pkg/cmd/cli/cmd/rsync/rsync.go

```diff
@@ -229,7 +236,89 @@ func (o *RsyncOptions) Validate() error {
 
 // RunRsync copies files from source to destination
 func (o *RsyncOptions) RunRsync() error {
-	return o.Strategy.Copy(o.Source, o.Destination, o.Out, o.ErrOut)
+	if err := o.Strategy.Copy(o.Source, o.Destination, o.Out, o.ErrOut); err != nil {
+		return err
+	}
+
+	if !o.Watch {
+		return nil
+	}
+	return o.WatchAndSync()
+}
+
+// WatchAndSync sets up a recursive filesystem watch on the sync path
+// and invokes rsync each time the path changes
+func (o RsyncOptions) WatchAndSync() error {
+
+	// these variables must be accessed while holding the changeLock
+	// mutex as they are shared between goroutines to communicate
+	// sync state/events.
+	var (
+		changeLock sync.Mutex
+		dirty      bool
+		lastChange time.Time
+		watchError error
+	)
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("error setting up filesystem watcher: %v", err)
+	}
+	defer watcher.Close()
+
+	go func() {
+		for {
+			select {
+			case event := <-watcher.Events:
+				changeLock.Lock()
+				glog.V(5).Infof("filesystem watch event: %s", event)
+				lastChange = time.Now()
+				dirty = true
+				if event.Op&fsnotify.Remove == fsnotify.Remove {
+					if e := watcher.Remove(event.Name); e != nil {
```

Comment:
start by copying a nested directory structure into the watched dir:

```
I0401 19:43:05.990199   25272 rsync.go:274] filesystem watch event: "./a": CREATE
I0401 19:43:05.990394   25272 fsnotification.go:30] adding watch on path ./a
I0401 19:43:05.990420   25272 fsnotification.go:30] adding watch on path a/b
I0401 19:43:05.990434   25272 fsnotification.go:30] adding watch on path a/b/c
I0401 19:43:05.990452   25272 rsync.go:274] filesystem watch event: "./a": CHMOD
I0401 19:43:05.990542   25272 fsnotification.go:30] adding watch on path ./a
I0401 19:43:05.990552   25272 fsnotification.go:30] adding watch on path a/b
I0401 19:43:05.990558   25272 fsnotification.go:30] adding watch on path a/b/c
I0401 19:43:09.391419   25272 rsync.go:314] Synchronizing filesystem changes...
I0401 19:43:09.391455   25272 copyrsync.go:76] Copying files with rsync
I0401 19:43:09.391478   25272 execlocal.go:19] Local executor running command: rsync -a --blocking-io --omit-dir-times --numeric-ids -v --delete -e oc rsh --loglevel=5 . jenkins-1-9wih5:/var/lib/jenkins/t
sending incremental file list
a/
a/a1
a/b/
a/b/b1
a/b/c/
a/b/c/c1

sent 258 bytes  received 100 bytes  716.00 bytes/sec
total size is 0  speedup is 0.00
I0401 19:43:09.456378   25537 exec.go:183] defaulting container name to jenkins
I0401 19:43:09.595606   25272 rsync.go:319] Done.
```

then `rm -rf a`
```
I0401 19:43:11.210944   25272 rsync.go:274] filesystem watch event: "a/b/b1": REMOVE
W0401 19:43:11.211004   25272 rsync.go:279] error removing watch for a/b/b1: can't remove non-existent inotify watch for: a/b/b1
I0401 19:43:11.211027   25272 rsync.go:274] filesystem watch event: "a/b/c/c1": REMOVE
W0401 19:43:11.211043   25272 rsync.go:279] error removing watch for a/b/c/c1: can't remove non-existent inotify watch for: a/b/c/c1
I0401 19:43:11.211074   25272 rsync.go:274] filesystem watch event: "a/b/c": REMOVE
W0401 19:43:11.211118   25272 rsync.go:279] error removing watch for a/b/c: can't remove non-existent inotify watch for: a/b/c
I0401 19:43:11.211134   25272 rsync.go:274] filesystem watch event: "a/b/c": REMOVE
W0401 19:43:11.211150   25272 rsync.go:279] error removing watch for a/b/c: can't remove non-existent inotify watch for: a/b/c
I0401 19:43:11.211177   25272 rsync.go:274] filesystem watch event: "a/b": REMOVE
W0401 19:43:11.211198   25272 rsync.go:279] error removing watch for a/b: can't remove non-existent inotify watch for: a/b
I0401 19:43:11.211208   25272 rsync.go:274] filesystem watch event: "a/b": REMOVE
W0401 19:43:11.211220   25272 rsync.go:279] error removing watch for a/b: can't remove non-existent inotify watch for: a/b
I0401 19:43:11.211233   25272 rsync.go:274] filesystem watch event: "a/a1": REMOVE
W0401 19:43:11.211244   25272 rsync.go:279] error removing watch for a/a1: can't remove non-existent inotify watch for: a/a1
I0401 19:43:11.211253   25272 rsync.go:274] filesystem watch event: "a": REMOVE
W0401 19:43:11.211265   25272 rsync.go:279] error removing watch for a: invalid argument
I0401 19:43:11.211401   25272 rsync.go:274] filesystem watch event: "./a": REMOVE
W0401 19:43:11.211415   25272 rsync.go:279] error removing watch for ./a: can't remove non-existent inotify watch for: a
I0401 19:43:13.391481   25272 rsync.go:314] Synchronizing filesystem changes...
I0401 19:43:13.391513   25272 copyrsync.go:76] Copying files with rsync
I0401 19:43:13.391539   25272 execlocal.go:19] Local executor running command: rsync -a --blocking-io --omit-dir-times --numeric-ids -v --delete -e oc rsh --loglevel=5 . jenkins-1-9wih5:/var/lib/jenkins/t
sending incremental file list
deleting a/b/c/c1
deleting a/b/c/
deleting a/b/b1
deleting a/b/
deleting a/a1
deleting a/

sent 40 bytes  received 65 bytes  210.00 bytes/sec
total size is 0  speedup is 0.00
I0401 19:43:13.484074   25568 exec.go:183] defaulting container name to jenkins
I0401 19:43:13.633249   25272 rsync.go:319] Done.
```

notice that not only do we get events for all the deletes, but something has actually already implicitly removed the watches.  so if anything we just need to not send that warning.